src/memory.rs

```diff
@@ -98,14 +98,18 @@ impl MemoryInstance {
 	fn new(limits: ResizableLimits) -> Result<Self, Error> {
 		check_limits(&limits)?;
 
-		let maximum_size = match limits.maximum() {
-			Some(maximum_pages) if maximum_pages > LINEAR_MEMORY_MAX_PAGES =>
-				return Err(Error::Memory(format!("maximum memory size must be at most {} pages", LINEAR_MEMORY_MAX_PAGES))),
-			Some(maximum_pages) => maximum_pages.saturating_mul(LINEAR_MEMORY_PAGE_SIZE),
-			None => u32::MAX,
-		};
-		let initial_size = calculate_memory_size(0, limits.initial(), maximum_size)
-			.ok_or_else(|| Error::Memory(format!("initial memory size must be at most {} pages", LINEAR_MEMORY_MAX_PAGES)))?;
+		let initial_pages = limits.initial();
+		let maximum_pages = limits.maximum().unwrap_or(LINEAR_MEMORY_MAX_PAGES);
+
+		if initial_pages > LINEAR_MEMORY_MAX_PAGES {
+			return Err(Error::Memory(format!("initial memory size must be at most {} pages", LINEAR_MEMORY_MAX_PAGES)));
+		}
+		if maximum_pages > LINEAR_MEMORY_MAX_PAGES {
+			return Err(Error::Memory(format!("maximum memory size must be at most {} pages", LINEAR_MEMORY_MAX_PAGES)));
+		}
+
+		let maximum_size = maximum_pages.saturating_mul(LINEAR_MEMORY_PAGE_SIZE);
+		let initial_size = calculate_memory_size(0, limits.initial(), maximum_size).expect("ok");
```

Comment:
expect should describe why it can't ever fail

Reply:
nice catch!